File: strategy/cpp/strategy.cc
#include "strategy/cpp/strategy.h"
#include <string>
#include <memory>

PaymentContext::PaymentContext(
    const std::string &sender_, const std::string &reciver_, const int money_
):sender(sender_), reciver(reciver_), money(money_) {}

Payment::Payment(const std::string &sender_,
    const std::string &reciver_,
    const int money_,
    std::shared_ptr<PaymentStrategy> strategy_) {
    context = std::make_shared<PaymentContext>(sender_, reciver_, money_);
    strategy = strategy_;
}

std::string Payment::Pay() const {
    return strategy->Pay(context);
}

std::string CacheStrategy::Pay(std::shared_ptr<PaymentContext> context) const {
    return "Pay $" + std::to_string(context->money) + " to " + context->reciver + " by cash";
}

std::string BanckStrategy::Pay(std::shared_ptr<PaymentContext> context) const {
    return "Pay $" + std::to_string(context->money) + " to " + context->reciver + " by bank account " + context->sender;
}

File: strategy/strategy.go
package strategy

import "fmt"

type Payment struct {
	context  *paymentContext
	strategy payStrategy
}

func NewPayment(receiver, sender string, money int, strategy payStrategy) *Payment {
	return &Payment{
		context: &paymentContext{
			sender:   sender,
			receiver: receiver,
			money:    money,
		},
		strategy: strategy,
	}
}

type paymentContext struct {
	sender   string
	receiver string
	money    int
}

type payStrategy interface {
	Pay(*paymentContext) string
}

func (p *Payment) Pay() string {
	return p.strategy.Pay(p.context)
}

type Cash struct{}

func (*Cash) Pay(ctx *paymentContext) string {
	return fmt.Sprintf("Pay $%d to %s by cash", ctx.money, ctx.receiver)
}

type Bank struct{}

func (*Bank) Pay(ctx *paymentContext) string {
	return fmt.Sprintf("Pay $%d to %s by bank account %s", ctx.money, ctx.receiver, ctx.sender)
}
